Support "primitiveVariables" argument to distributePoints

danieldresser-ie opened the pull request:
... and some odds and ends that came up along the way.  I think based on our previous conversations, it's probably fairly clear where most of this came from.

The final commit can safely be dropped as out of scope - it isn't necessary.  Though I do consider it pretty wrong that we potentially waste a lot of time generating points that will necessarily be rejected based on the densities of all the vertices of the current triangle.  The test case shows a factor of 2 speedup, but that could be a lot more dramatic if you wanted to generate millions of points one small area of a large mesh, so you used an extremely high global density, and then had a primitive variable that is zero for most of the mesh ... this case doesn't seem that implausible if you were using Seeds for environment layout.  I've left a test failure in though, illustrating that shifting the test inside PointsDistribution changes the behaviour for points that are exactly on the threshold of being accepted ( because distributePoints uses an inclusive comparison, but PointsDistribution seems to implicitly use an exclusive comparison ... floating point error could result in points being rejected that were previously on the threshold anyway ).  If we're planning to alter the algorithm for PointsDistribution at some point, I guess that would be the time to merge this.  That also might be a good time to get rid of the duplicate points that PointsDistribution generates, which I noticed looking at the test data.

johnhaddon commented on src/IECoreScene/MeshAlgoDistributePoints.cpp:
```diff
@@ -134,39 +139,93 @@ T triangleInterpolatedPrimVarValue(
+	{
+		processedMesh->variables["P"] = pIt->second;
+
+		if( position != "P" && StringAlgo::match( position, primitiveVariables ) )
```
`matchMultiple()`

johnhaddon commented on src/IECoreScene/MeshAlgoDistributePoints.cpp:
```diff
@@ -189,15 +248,29 @@ void processInputs(
+			densityVar = densityVarIt->second;
+		}
+
+		if( !StringAlgo::match( densityMask, primitiveVariables ) )
```
`matchMultiple()`.

johnhaddon commented on src/IECoreScene/MeshAlgoDistributePoints.cpp:
```diff
@@ -134,39 +139,93 @@ T triangleInterpolatedPrimVarValue(
+	{
+		if( var.first == position )
+		{
+			// \todo - is it a problem that this P that we are setting up solely to be passed in
+			// triangulate ( where it isn't used ), may not have a Point interpretation set correctly?
```
> (where it isn't used)

I don't understand this comment - it _is_ used in `triangulate()`, otherwise we wouldn't be passing it would we?

johnhaddon replied:
I'm not really following all this juggling of `P` and `position` - it seems _painful_. Wasn't it broken before anyway? Since we don't use it in Gaffer, I'd be perfectly happy to ignore it if it simplifies the code. Maybe we can just throw if `position != "P"` and then remove the parameter in 10.5?

danieldresser-ie replied:
Yeah, it is fairly painful, and that's after going through a couple of rearrangements that felt even worse. I'm happy to remove this feature if it feels unnecessary. It feels like a bit of a shame to not remove the parameter already in the new signature, since we already need to add a new signature and deprecate the old one - but unfortunately since MatchPattern and string are actually the same type, and the place for "primitiveVariables" in the arguments really does seem to be the same as where "position" used to be, I guess there really isn't a good way to differentiate them.  Throwing is probably good enough for now ... I guess we don't really much about non-Gaffer uses anyway.

johnhaddon replied:
We could perhaps have called the new function `scatterPoints()` instead. But actually I think we have completely misinterpreted the intention of the `position` argument, and our "fix" for it was wrong.

If you go back to the [origins](https://github.com/ImageEngine/cortex/commit/eaa8243269ef23262f3f20ccf84db337f9a44cc2#diff-a27f3120a622be0b9a2f44250a807354c8c3c60e0e930126ca8adfc97235c764) of this code, you'll see that it was ported from an op, and that in the op the parameter was called `pRefPrimVarName`. That explains why `position` was only being used to calculate the face areas, and _not_ being used to feed the output points. It's only purpose is to get a stable distribution on a deforming mesh - like you'd need for fur follicles on a creature, for instance. So we definitely need to keep the parameter (although we should probably rename it), and we definitely need to keep its original meaning. Hopefully that also means some of the juggling can be removed?

danieldresser-ie replied:
Oh, yeah, I was pretty unsure about changing the meaning of it when I asked you, but I guess we were both mislead by the name + lack of tests. And yeah, it should be more straightforward to make it work that way. Does "refPosition" sound good? Or "referencePosition"?

johnhaddon replied:
Yeah, either of those sounds good - a test case to guide our future selves would be handy too.

johnhaddon commented on src/IECoreScene/MeshAlgoDistributePoints.cpp:
```diff
@@ -134,39 +139,93 @@ T triangleInterpolatedPrimVarValue(
+		pIt->second.data->typeId() == V3fVectorDataTypeId
 	)
+	{
+		processedMesh->variables["P"] = pIt->second;
```
What is this assignment for? Since `pIt == find( "P" )` then aren't we just assigning the same data again?

johnhaddon commented on src/IECoreScene/MeshAlgoDistributePoints.cpp:
```diff
@@ -313,8 +313,38 @@ void distributePointsInTriangle(
+		// cannot increase the number of points generated. Maybe this is valuable as a safety mechanism?
+		// We could just delete this line, and then densities greater than 1 would work how you would expect.
+		maxDensity = std::min( 1.0f, maxDensity );
+		//maxDensity = std::max( 1.0f, maxDensity );
```
Dead code

johnhaddon reviewed (commented):
All pretty minor comments from me - see inline. In addition to those, I did spot one difference in behaviour with this new version compared to the old. A negative density provided by a constant primvar is causing points to be generated, whereas before it would prevent them. Repro below...

<details>
```
import Gaffer
import GafferScene
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 2, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 3, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Plane2"] = GafferScene.Plane( "Plane2" )
parent.addChild( __children["Plane2"] )
__children["Plane2"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Scatter2"] = GafferScene.Scatter( "Scatter2" )
parent.addChild( __children["Scatter2"] )
__children["Scatter2"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["PathFilter6"] = GafferScene.PathFilter( "PathFilter6" )
parent.addChild( __children["PathFilter6"] )
__children["PathFilter6"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["PrimitiveVariables1"] = GafferScene.PrimitiveVariables( "PrimitiveVariables1" )
parent.addChild( __children["PrimitiveVariables1"] )
__children["PrimitiveVariables1"]["primitiveVariables"].addChild( Gaffer.NameValuePlug( "", Gaffer.FloatPlug( "value", defaultValue = 0.0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member1", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["PrimitiveVariables1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["DeleteObject3"] = GafferScene.DeleteObject( "DeleteObject3" )
parent.addChild( __children["DeleteObject3"] )
__children["DeleteObject3"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Plane2"]["__uiPosition"].setValue( imath.V2f( 136.164764, 25.258091 ) )
__children["Scatter2"]["in"].setInput( __children["PrimitiveVariables1"]["out"] )
__children["Scatter2"]["filter"].setInput( __children["PathFilter6"]["out"] )
__children["Scatter2"]["name"].setValue( 'scatter' )
__children["Scatter2"]["density"].setValue( 100.0 )
__children["Scatter2"]["densityPrimitiveVariable"].setValue( 'density' )
__children["Scatter2"]["__uiPosition"].setValue( imath.V2f( 136.164764, 3.24538898 ) )
__children["PathFilter6"]["paths"].setValue( IECore.StringVectorData( [ '/plane' ] ) )
__children["PathFilter6"]["__uiPosition"].setValue( imath.V2f( 157.311035, 20.4606438 ) )
__children["PrimitiveVariables1"]["in"].setInput( __children["Plane2"]["out"] )
__children["PrimitiveVariables1"]["filter"].setInput( __children["PathFilter6"]["out"] )
__children["PrimitiveVariables1"]["primitiveVariables"]["member1"]["name"].setValue( 'density' )
__children["PrimitiveVariables1"]["primitiveVariables"]["member1"]["value"].setValue( -0.10000000149011612 )
__children["PrimitiveVariables1"]["__uiPosition"].setValue( imath.V2f( 134.726532, 14.2521162 ) )
__children["DeleteObject3"]["in"].setInput( __children["Scatter2"]["out"] )
__children["DeleteObject3"]["filter"].setInput( __children["PathFilter6"]["out"] )
__children["DeleteObject3"]["__uiPosition"].setValue( imath.V2f( 136.165604, -4.91867352 ) )


del __children
```
</details>

danieldresser-ie commented on src/IECoreScene/MeshAlgoDistributePoints.cpp:
```diff
@@ -184,6 +188,16 @@ struct Generator
+				if( ! ( approxCandidatePoints <= maxCandidatePoints ) )
+				{
+					std::string e = boost::str( boost::format(
+						"MeshAlgo::distributePoints : Cannot generate more than %i candidate points per polygon. Trying to generate %i. There are circumstances where the output would be reasonable, but this happens during processing due to a polygon with a large area in 3D space which is extremely thin and diagonal in UV space, in which case you may need to clean your UVs. Alternatively, maybe you really want to p ...
```
If the area is small but not thin, then we will only visit wang tiles that overlap the tiny area.  There is still a cost in visiting the upper levels of the hierarchy, but this is basically the same as for any highly subdivided mesh - we're never particularly efficient, but you don't get the really nasty case of thin and diagonal where we end up visiting huge numbers of leaf tiles that don't even overlap the target triangle at all.  If the area is small and thin but not diagonal, then that is moderately bad, though definitely less bad than thin and diagonal - I think basically we're not visiting tiles that don't overlap the triangle at all, but a thin triangle can still span a bunch of leaf tiles.  I'm up for simply omitting "and diagonal", since any cleaning process would probably not be specific to diagonals, but I think "thin" is important to what triggers a dramatic slowdown.

danieldresser-ie replied:
Yeah, that alternate PointDistribution call could be interesting - I'm not sure it would be worth it for the thin triangle special case, but if you were generating a very large number of points on a fairly low polygon mesh, you would theoretically save half the points that need testing. Testing the tile bounds against the triangle would have a cost though - maybe you would need a metric where if there are less than <someNumber> points left to be generated, then it's worth testing if we can skip the current tile.

danieldresser-ie commented:
I think I've addressed all feedback, along with some other stuff that came up along the way ( like duplicated points and low precision barycentric calculations ).

It took some shuffling to try and make sure the commits dealt with an understandable piece of things at each step, but also kept the tests passing at each step ... you might need to skim back through all the commits ( maybe I should have structured this better ... but I think it would have been a total mess if I was trying not to squash changes ).

Of particular note is the final commit where I remove the limit that density primvars do nothing if set over 1 - this is a change in behaviour, but I think a worthwhile one? Easy to drop it if you feel otherwise though.

danieldresser-ie commented:
Retargeted to main, as per conversation.

danieldresser-ie commented:
Fixed bad asserts ( remaining test failure is an unrelated timing failure ).
